### PR TITLE
`node.config.php` bugfixings

### PR DESCRIPTION
**Pull request by nupplaphil**
This is a general bugfixing of the `Config` behavior. Most of them introduced by #12593 , but not all :-)
f.e. currently, deleting "just" resets the config to the static default value and doesn't "really" deletes the value.

I tested the whole upgrade path from `stable` to `develop` a few times.
This should solve #12616 and https://github.com/friendica/friendica/pull/12593#issuecomment-1372598717

I added unit tests for edge-cases too..

Now it will run at my node over the next ~12hours and I will check my logs again.

### CODE REVIEW

File: src/Core/Config/Util/ConfigFileManager.php [hunk 1]
```diff
@@ -177,49 +182,83 @@ private function loadDataConfig(Cache $config)
...
+			if (is_array($dataArray)) {
+				$configCache->load($dataArray, Cache::SOURCE_DATA);
 			}
-
-			$config->load($dataArray, Cache::SOURCE_DATA);
 		}
 	}
 
 	/**
 	 * Saves overridden config entries back into the data.config.phpR
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileManager.php]:
```suggestion
	 * Saves overridden config entries back into the data.config.php
```

File: src/Core/Config/Util/ConfigFileManager.php [hunk 2]
```diff
@@ -177,49 +182,83 @@ private function loadDataConfig(Cache $config)
...
-		if (!$encodedData) {
-			throw new ConfigFileException('config source cannot get encoded');
+		if (file_exists($filename)) {
+			$fileExists = true;
+		} else {
+			$fileExists = false;
 		}
 
-		$configStream = fopen($this->configDir . '/' . self::CONFIG_DATA_FILE, 'w');
+		$configStream = fopen($filename, 'c+');
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileManager.php:247]: Please vary the flag depending on `$fileExists` to replace some of the following convolution.

File: tests/src/Core/Config/Cache/CacheTest.php [hunk 3]
```diff
@@ -67,6 +67,8 @@ public function testLoadConfigArray($data)
 		$configCache = new Cache();
 		$configCache->load($data);
 
+		print_r($configCache);
+
```
Comment (MrPetovan) [inline tests/src/Core/Config/Cache/CacheTest.php]: Please remove debug when you're done.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Core/Config/Util/ConfigFileManager.php (hunk 2 above)
Comment (nupplaphil) [reply src/Core/Config/Util/ConfigFileManager.php:247]: I read about a very very small chance for a race condition during this check, so two `save()` will try both a `w+` instead a `w+` and `r+` lock. That's why I let it the OS decide what to do :-)

Comment (MrPetovan) [reply src/Core/Config/Util/ConfigFileManager.php:247]: Ok, I feel like this block deserves a lengthy explanation in a multi-line comment.

Comment (nupplaphil) [reply src/Core/Config/Util/ConfigFileManager.php:247]: Aaaand here's the doc :-)

But plz have a look at my english grammer/syntax, because it's hard for me as non-native speaker (and tbh I'm frequently writing english only for Friendica *gg*)

File: src/Core/Config/Util/ConfigFileManager.php [hunk 4]
```diff
@@ -144,82 +149,153 @@ private function loadStaticConfig(string $name): array
...
 			}
 
-			fclose($configStream);
-
+			/**
+			 * Evaluate the fetched content
+			 *
+			 * @note
+			 * Because `eval()` directly evaluates PHP content, we need to "close" the expected PHP content again with
+			 * the prefixed "?>". Now we're in plain HTML again and can evaluate any PHP file :-)
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileManager.php]:
```suggestion
			 * the prefixed "?>". Now we're in plain text again and can evaluate any PHP file :-)
```

File: src/Core/Config/Util/ConfigFileManager.php [hunk 5]
```diff
@@ -144,82 +149,153 @@ private function loadStaticConfig(string $name): array
...
+				if ($fileExists) {
+					// When reading the config file too fast, we get a wrong filesize, "clearstatcache" prevents that
+					clearstatcache(true, $filename);
+					$content = fread($configStream, filesize($filename));
+					// Event truncating the whole content wouldn't automatically rewind the stream,
+					// so we need to do it manually
+					rewind($configStream);
+					if (!$content) {
+						throw new ConfigFileException(sprintf('Cannot read file %s', $filename));
+					}
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileManager.php]: Is it meaningful to rewind the pointer before checking for content?
```suggestion
					if (!$content) {
						throw new ConfigFileException(sprintf('Cannot read file %s', $filename));
					}

					// Event truncating the whole content wouldn't automatically rewind the stream,
					// so we need to do it manually
					rewind($configStream);
```

File: src/Core/Config/Util/ConfigFileManager.php [hunk 6]
```diff
@@ -144,82 +149,153 @@ private function loadStaticConfig(string $name): array
...
+		 *
+		 */
+		$configStream = fopen($filename, 'c+');
+
+		try {
+			// We do want an exclusive lock, so we wait until every LOCK_SH (config reading) is unlocked
+			if (flock($configStream, LOCK_EX)) {
+
+				/**
+				 * If the file exists, read the whole file again
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileManager.php]:
```suggestion
				 * If the file exists, we read the whole file again to avoid a race condition with concurrent threads that could have modified the file between the first config read of this thread and now
```

File: src/Core/Config/Util/ConfigFileManager.php [hunk 7]
```diff
@@ -144,82 +149,153 @@ private function loadStaticConfig(string $name): array
...
+				 * Since we're currently exclusive locked, no other process can now change the config again
+				 */
+				if ($fileExists) {
+					// When reading the config file too fast, we get a wrong filesize, "clearstatcache" prevents that
+					clearstatcache(true, $filename);
+					$content = fread($configStream, filesize($filename));
+					// Event truncating the whole content wouldn't automatically rewind the stream,
+					// so we need to do it manually
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileManager.php]:
```suggestion
					// Even truncating the file doesn't rewind the stream after the full read, so we need to do it explicitly
```

File: src/Core/Config/Util/ConfigFileManager.php [hunk 8]
```diff
@@ -144,82 +149,153 @@ private function loadStaticConfig(string $name): array
...
+						$fileConfigCache = new Cache();
+						$fileConfigCache->load($dataArray, Cache::SOURCE_DATA);
+						$configCache = $fileConfigCache->merge($configCache);
+					}
+				}
+
+				// Only SOURCE_DATA is wanted, the rest isn't part of the node.config.php file
+				$data = $configCache->getDataBySource(Cache::SOURCE_DATA);
+
+				$encodedData = ConfigFileTransformer::encode($data);
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileManager.php]:
```suggestion
```

File: src/Core/Config/Util/ConfigFileTransformer.php [hunk 9]
```diff
@@ -34,6 +34,12 @@ public static function encode(array $data): string
 		$categories = array_keys($data);
 
 		foreach ($categories as $category) {
+
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileTransformer.php]:
```suggestion
```

File: src/Core/Config/ValueObject/Cache.php [hunk 10]
```diff
@@ -159,40 +160,49 @@ public function getDataBySource(int $source): array
...
+		if ((isset($this->source[$cat][$key]) && $source < $this->source[$cat][$key]) ||
+			(is_int($this->source[$cat] ?? null) && $source < $this->source[$cat])) {
 			return false;
 		}
 
 		if ($this->hidePasswordOutput &&
 			$key == 'password' &&
 			is_string($value)) {
-			$this->config[$cat][$key] = new HiddenString((string)$value);
+			$this->setCatKeyValueSource($cat, $key, new HiddenString((string)$value), $source);
```
Comment (MrPetovan) [inline src/Core/Config/ValueObject/Cache.php]: `$value` already is a string per condition.
```suggestion
			$this->setCatKeyValueSource($cat, $key, new HiddenString($value), $source);
```

File: src/Core/Update.php [hunk 11]
```diff
@@ -141,7 +161,15 @@ public static function run(string $basePath, bool $force = false, bool $override
...
-					$retryBuild = DI::config()->get('system', 'build', null);
+					$retryBuild = DI::config()->get('system', 'build');
+					// legacy option - check if there's something in the Config table
+					if (DBStructure::existsTable('config')) {
+						$dbConfig = DBA::selectFirst('config', ['v'], ['cat' => 'system', 'k' => 'build']);
+						if (!empty($dbConfig)) {
+							$retryBuild = $dbConfig['v'];
+						}
+					}
+
```
Comment (MrPetovan) [inline src/Core/Update.php]: At this point we have a value, ensured in line 144.
```suggestion
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: bin/daemon.php [hunk 12]
```diff
@@ -63,7 +64,9 @@
 
 DI::init($dice);
 \Friendica\Core\Logger\Handler\ErrorHandler::register($dice->create(\Psr\Log\LoggerInterface::class));
-$a = DI::app();
+
+// Check the database structure and possibly fixes it
+Update::check(DI::basePath(), true, DI::mode());
```
Comment (annando) [inline bin/daemon.php]: That check will only be done on restarting the daemon, If you want to enable the check during normal processing, you have to place it inside the loop, possibly at the place where the cron call is executed.

File: index.php [hunk 13]
```diff
@@ -36,6 +36,9 @@
 
 \Friendica\Core\Logger\Handler\ErrorHandler::register($dice->create(\Psr\Log\LoggerInterface::class));
 
+// Check the database structure and possibly fixes it
+\Friendica\Core\Update::check(\Friendica\DI::basePath(), true, \Friendica\DI::mode());
```
Comment (annando) [inline index.php]: We are already having this call here: https://github.com/friendica/friendica/blob/develop/src/App.php#L662

Does this means that it will be called now twice?

Also: setting the second parameter to true means that the update will be called directly via the frontend process. Frontend processes normally have a maximum lifespan of maybe two minutes. This means that longer update processes will be interrupted and would leave the system in some ugly intermediate state.

File: bin/worker.php [hunk 14]
```diff
@@ -58,12 +58,11 @@
 
 DI::init($dice);
 \Friendica\Core\Logger\Handler\ErrorHandler::register($dice->create(\Psr\Log\LoggerInterface::class));
-$a = DI::app();
 
 DI::mode()->setExecutor(Mode::WORKER);
 
 // Check the database structure and possibly fixes it
-Update::check($a->getBasePath(), true, DI::mode());
+Update::check(DI::basePath(), true, DI::mode());
```
Comment (annando) [inline bin/worker.php]: I just had a look at the code. It seems as if neither the first nor the third parameter is in use.

File: index.php (hunk 13 above)
Comment (nupplaphil) [reply index.php]: removed again, you're right!

File: bin/daemon.php (hunk 12 above)
Comment (nupplaphil) [reply bin/daemon.php]: Because there's already a check inside the worker itself, we don't need it at the daemon too .. Seems redundant

Comment (MrPetovan) [reply bin/daemon.php]: What worker? `bin/worker.php`? It isn't used by the daemon.

Comment (nupplaphil) [reply bin/daemon.php]: oh, my mistake sry

**Review by MrPetovan — APPROVED** [review]

File: src/Core/Update.php (hunk 11 above)
Comment (nupplaphil) [reply src/Core/Update.php]: This doesn't work because the next `DI::config()->get('system', 'build') at line 164 retrieves the wrong build number again, so I had to check it here again

**Comment by nupplaphil** [comment]
Now the Update process really works.

My test:
- checkout stable (1502)
- start daemon
- check log for at least one execution (cron start / end)
- checkout this PR branch (1509)
- Wait for next interval
- node.config.php got completely migrated (build is now 1509)

**Comment by annando** [comment]
To make it work for daemon systems as well, I suggest to apply some change to the worker, so that jobs with the priority `PRIORITY_CRITICAL` are executed anyway, independent on the daemon status. This ensures the execution.

**Comment by nupplaphil** [comment]
> To make it work for daemon systems as well, I suggest to apply some change to the worker, so that jobs with the priority `PRIORITY_CRITICAL` are executed anyway, independent on the daemon status. This ensures the execution.

tbh I'm not entirely sure what you mean. Where do you see the changes? In daemon mode, how can a worker can work without a proper daemon status??

**Comment by MrPetovan** [comment]
Awaiting @annando ’s answer before merging.

**Comment by annando** [comment]
See https://github.com/friendica/friendica/blob/develop/src/Core/Worker.php#L1319

There we quit when we are in the daemon mode. On that priority we could skip this, which would spawn a worker.

**Comment by nupplaphil** [comment]
@annando - I thought that's enough, isn't it?

Because the worker already prioritize critical tasks, see https://github.com/friendica/friendica/blob/develop/src/Core/Worker.php#L952-L954

**Review by MrPetovan — APPROVED** [review]

**Comment by annando** [comment]
Sadly this hadn't solved my issue ...
